filters/include/pcl/filters/experimental/voxel_grid_label.h

```diff
@@ -0,0 +1,111 @@
+/*
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ *  Point Cloud Library (PCL) - www.pointclouds.org
+ *  Copyright (c) 2021-, Open Perception
+ *
+ *  All rights reserved
+ */
+
+#pragma once
+
+#include <pcl/filters/experimental/voxel_grid.h>
+#include <pcl/filters/filter.h>
+#include <pcl/point_types.h>
+
+#include <map>
+
+namespace pcl {
+namespace experimental {
+
+struct LabeledVoxel : public Voxel<PointXYZRGBL> {
+  using Voxel<PointXYZRGBL>::centroid_;
+  using Voxel<PointXYZRGBL>::downsample_all_data_;
+  using Voxel<PointXYZRGBL>::num_pt_;
+
+  LabeledVoxel(const bool downsample_all_data)
+  : Voxel<PointXYZRGBL>(downsample_all_data)
+  {}
+
+  inline void
+  add(const PointXYZRGBL& pt)
+  {
+    num_pt_++;
+    if (downsample_all_data_) {
+      centroid_.all_fields.add(pt);
+    }
+    else {
+      centroid_.xyz += pt.getArray4fMap();
+      labels_[pt.label]++;
+    }
+  }
+
+  inline PointXYZRGBL
+  get() const
+  {
+    PointXYZRGBL pt;
+    if (downsample_all_data_)
+      centroid_.all_fields.get(pt);
+    else {
+      pt.getArray4fMap() = centroid_.xyz / num_pt_;
+      std::size_t max = 0;
+      for (const auto& label : labels_) {
+        if (label.second > max) {
+          max = label.second;
+          pt.label = label.first;
+        }
+      }
+    }
+
+    return pt;
+  }
+
+  inline void
+  clear()
+  {
+    this->clear();
+    if (!downsample_all_data_)
+      labels_.clear();
+  }
+
+protected:
+  std::map<std::uint32_t, std::size_t> labels_;
+};
+
+/**
+ * \brief LabeledVoxelStruct defines the transformation operations and the voxel grid of
+ * VoxelGridLabel filter
+ * \ingroup filters
+ */
+class LabeledVoxelStruct : public VoxelStruct<Voxel<PointXYZRGBL>, PointXYZRGBL> {
+public:
+  using VoxelStruct<Voxel<PointXYZRGBL>, PointXYZRGBL>::filter_name_;
+
+  /** \brief Constructor. */
+  LabeledVoxelStruct() { filter_name_ = "VoxelGridLabel"; }
+
+  /** \brief Empty destructor. */
+  ~LabeledVoxelStruct() {}
+
+  /** \brief Set up the voxel grid variables needed for filtering.
+   *  \param[in] transform_filter pointer to the TransformFilter object
+   */
+  bool
+  setUp(CartesianFilter<pcl::Filter, LabeledVoxelStruct>* castesian_filter)
```

Comment:
Instead of pointer, if we use a reference, we can be ensured that this will never be null

Reply:
Done